src/sasl/cyrsasl_external.erl

```diff
@@ -66,73 +76,150 @@ mech_step(#state{creds = Creds}, User) ->
+        {error, Error} when T =:= [] ->
+            {error, Error};
+        {error, _} ->
+            verification_loop(T, Creds);
+        {ok, Name} ->
+            {ok, Name}
     end.
 
-maybe_use_common_name(_, User, Server, allow_just_user_identity) ->
-    verify_server(User, Server);
-maybe_use_common_name(CommonName, User, Server, use_common_name) ->
-    case verify_server(User, Server) of
-        {ok, CommonName} ->
+verify_creds(standard, Creds) ->
+    standard_verification(Creds);
+verify_creds(common_name, Creds) ->
+    common_name_verification(Creds);
+verify_creds(auth_id, Creds) ->
+    auth_id_verification(Creds);
+verify_creds({mod, Mod}, Creds) ->
+    custom_verification(Mod, Creds).
+
+
+standard_verification(Creds) ->
+    Server = mongoose_credentials:lserver(Creds),
+    XmppAddrs = get_credentials(Creds, xmpp_addresses),
+    AuthId = get_credentials(Creds, auth_id),
+    XmppAddr = case {XmppAddrs, AuthId} of
+                   {[OneXmppAddr], undefined} ->
+                       OneXmppAddr;
+                   {[_], _} ->
+                       {error, <<"invalid-authzid">>};
+                   {[_, _ | _], undefined} ->
+                       {error, <<"invalid-authzid">>};
+                   _ ->
+                       case lists:member(AuthId, XmppAddrs) of
+                           true ->
+                               AuthId;
+                           _ ->
+                               {error, <<"not-authorized">>}
+                       end
+               end,
+    verify_server(XmppAddr, Server).
+
+common_name_verification(Creds) ->
+    Server = mongoose_credentials:lserver(Creds),
+    AuthId = get_credentials(Creds, auth_id),
+    CommonName = get_credentials(Creds, common_name),
+    case AuthId of
+        undefined when is_binary(CommonName) ->
             {ok, CommonName};
         _ ->
-            {error, <<"not-authorized">>}
-    end;
-maybe_use_common_name(_, _, _, standard) ->
-    {error, <<"not-authorized">>}.
-
-get_common_name(Cert) ->
-    case cert_utils:get_common_name(Cert) of
-        error -> [];
-        CN -> [{common_name, CN}]
+            case verify_server(AuthId, Server) of
+                {ok, CommonName} -> {ok, CommonName};
+                _ -> {error, <<"invalid-authzid">>}
+            end
     end.
 
-get_xmpp_addresses(Cert) ->
-    case cert_utils:get_xmpp_addresses(Cert) of
-        [] -> [{xmpp_addresses, []}];
-        XmmpAddresses -> [{xmpp_addresses, XmmpAddresses}]
+auth_id_verification(Creds) ->
+    Server = mongoose_credentials:lserver(Creds),
+    AuthId = get_credentials(Creds, auth_id),
+    verify_server(AuthId, Server).
+
+custom_verification(Module, Creds) ->
+    %% erlang:function_exported/3 returns false if module
+    %% is not loaded. assuming that all the modules are
+    %% loaded at startup. see start_all_modules/0 function
+    case erlang:function_exported(Module, verify_creds, 1) of
```

Comment:
I think in this case it's better to fail early and loud. If a module that's expected to implement certificate verification doesn't do it, then `cyrsasl_external` should refuse to start.

Reply:
sorry, cannot make this change as start/stop interfaces were removed from `cyrsasl` behaviour